### PR TITLE
Simplify ObjectMap interface

### PR DESCRIPTION
**Pull request by adrianbroher**
This PR simplifies the ObjectMap class interface:

* Use established naming conventions to mimic std containers.
  * NumObjects - size
  * Empty -> empty
  * Object -> get
  * FindObjects (with some query parameter) -> find
  * FindObjects (without query parameter) -> all
  * const_{begin,end} -> {begin, end} with const modifier
  * Insert -> insert
  * Remove -> erase
  * Clear -> clear
* Remove non-templates variants if they can be substituted by the template equivalent
  * begin
  * end
  * const_end
  * const_begin
  * FindObjects
  * Object
  * NumObjects
* Reimplement ObjectMap::find(const UniverseObjectVisitor&) as template member function to query only a subtype with an applied visitor.
* Remove internal ObjectMap helpers from class interface.
* Replace FindObjectIDs convenience functions with the corresponding all or find equivalent.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
I don't like the `Object` -> `at` renaming, as I find that name very unintuitive, and would much prefer something like `get`, and because I think `ObjectMap`'s function works differently from most (?) STL containers' `at` functions, in that if there is no such item in the container, it doesn't throw an exception, but rather returns a `null_ptr`.

File: GG/GG/ZList.h [hunk 1]
```diff
@@ -59,14 +59,14 @@ class GG_API ZList
     /** RenderOrderIterable can be iterated in back to front render order. */
     struct RenderOrderIterable {
         RenderOrderIterable(const std::list<std::shared_ptr<Wnd>>& list) :
-            m_list(list.crbegin(), list.crend())
+            m_list(list.rbegin(), list.rend())
```
Comment (geoffthemedio) [inline GG/GG/ZList.h]: seems unrelated to the subject of this pull request...

File: UI/OptionsWnd.cpp [hunk 2]
```diff
@@ -1279,7 +1279,7 @@ namespace {
 
         const std::regex dot { "\\.+" };
         const std::vector<std::string> nodes {
-            std::sregex_token_iterator(name.cbegin(), name.cend(), dot, -1),
+            std::sregex_token_iterator(name.begin(), name.end(), dot, -1),
```
Comment (geoffthemedio) [inline UI/OptionsWnd.cpp]: also seems unrelated

File: python/UniverseWrapper.cpp [hunk 3]
```diff
@@ -72,20 +72,14 @@ namespace {
...
-    std::vector<int>        PlanetIDs(const Universe& universe)
-    { return Objects().FindObjectIDs<Planet>(); }
-    std::vector<int>        ShipIDs(const Universe& universe)
-    { return Objects().FindObjectIDs<Ship>(); }
-    std::vector<int>        BuildingIDs(const Universe& universe)
-    { return Objects().FindObjectIDs<Building>(); }
+    template<typename T>
+    std::vector<int> ObjectIDs(const Universe& universe)
+    {
+        std::vector<int> result;
```
Comment (geoffthemedio) [inline python/UniverseWrapper.cpp]: probably worth reserving space before repeatedly pushing a known number of items into container

File: UI/OptionsWnd.cpp (hunk 2 above)
Comment (adrianbroher) [reply UI/OptionsWnd.cpp]: See #2730

File: GG/GG/ZList.h (hunk 1 above)
Comment (adrianbroher) [reply GG/GG/ZList.h]: See #2730

**Comment by adrianbroher** [comment]
Renamed `ObjectMap::at` to `ObjectMap::get`.

File: python/UniverseWrapper.cpp (hunk 3 above)
Comment (adrianbroher) [reply python/UniverseWrapper.cpp]: Done

File: combat/CombatSystem.cpp [hunk 4]
```diff
@@ -1303,7 +1301,7 @@ namespace {
         }
     };
 
-    std::vector<PartAttackInfo> GetWeapons(std::shared_ptr<UniverseObject>& attacker) {
+    std::vector<PartAttackInfo> GetWeapons(std::shared_ptr<UniverseObject> attacker) {
```
Comment (geoffthemedio) [inline combat/CombatSystem.cpp:1304]: Reason for passing by value and not reference? (Or better: const reference)

File: universe/Universe.cpp [hunk 5]
```diff
@@ -856,12 +856,17 @@ void Universe::UpdateMeterEstimatesImpl(const std::vector<int>& objects_vec, boo
...
         obj->BackPropagateMeters();
 }
 
 void Universe::BackPropagateObjectMeters()
-{ BackPropagateObjectMeters(m_objects.FindObjectIDs()); }
+{
+    std::vector<int> result;
+    for (const auto& obj : m_objects.all())
+        result.push_back(obj->ID());
+    BackPropagateObjectMeters(result);
```
Comment (geoffthemedio) [inline universe/Universe.cpp]: This version of `BackPropagateObjectMeters` should probably just do the `obj->BackPropagateMeters()` call directly on all the objects, rather than looping over objects to get their ids, push them into a vector, pass that vector into the overload, which then calls `find` with the vector of ids to get the objects again, in order to do the actual calls.

File: AI/AIInterface.cpp [hunk 6]
```diff
@@ -173,7 +173,7 @@ namespace AIInterface {
             // ownership to all, while remembering which planets this is done
             // to.
             universe.InhibitUniverseObjectSignals(true);
-            for (auto& planet : universe.Objects().FindObjects<Planet>()) {
+            for (auto& planet : universe.Objects().all<Planet>()) {
                  if (planet->Unowned()) {
                      unowned_planets.push_back(planet);
```
Comment (geoffthemedio) [inline AI/AIInterface.cpp:178]: not related to this pull request, but another case where reserving space before pushing into vector would be good...

File: UI/EncyclopediaDetailPanel.cpp [hunk 7]
```diff
@@ -297,7 +297,7 @@ namespace {
 
                 // occupied planets
                 std::vector<std::shared_ptr<const Planet>> species_occupied_planets;
```
Comment (geoffthemedio) [inline UI/EncyclopediaDetailPanel.cpp:299]: also here...

File: UI/EncyclopediaDetailPanel.cpp [hunk 8]
```diff
@@ -1671,10 +1671,9 @@ namespace {
 
         // Fleets
         std::vector<std::shared_ptr<const UniverseObject>> nonempty_empire_fleets;
-        for (auto& maybe_fleet : Objects().FindObjects(OwnedVisitor<Fleet>(empire_id))) {
-            if (auto fleet = std::dynamic_pointer_cast<const Fleet>(maybe_fleet))
-                if (!fleet->Empty())
-                    nonempty_empire_fleets.push_back(fleet);
+        for (const auto& fleet : Objects().find<Fleet>(OwnedVisitor<Fleet>(empire_id))) {
+            if (!fleet->Empty())
+                nonempty_empire_fleets.push_back(fleet);
```
Comment (geoffthemedio) [inline UI/EncyclopediaDetailPanel.cpp:1676]: and here... probably there are a lot that should be handled separately

File: universe/Universe.cpp (hunk 5 above)
Comment (adrianbroher) [reply universe/Universe.cpp]: The vector&lt;int&gt; variant is only used at this location so it can be dropped entirely.

File: AI/AIInterface.cpp (hunk 6 above)
Comment (adrianbroher) [reply AI/AIInterface.cpp:178]: There are multiple location where universe object id vector is not nessecary, but the size of the MR is already big enough as it is.

File: combat/CombatSystem.cpp (hunk 4 above)
Comment (adrianbroher) [reply combat/CombatSystem.cpp:1304]: GetWeapons ist not called with shared_ptr&lt;UniverseObject&gt; all the time, but shared_ptr&lt;Ship&gt;.

I'm also not sure it we are even using the shared_ptr as intended.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#S-resource
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-smartptrparam

File: universe/Universe.cpp (hunk 5 above)
Comment (adrianbroher) [reply universe/Universe.cpp]: Done

File: combat/CombatSystem.cpp (hunk 4 above)
Comment (geoffthemedio) [reply combat/CombatSystem.cpp:1304]: > GetWeapons ist not called with shared_ptr all the time, but shared_ptr.

?

Comment (adrianbroher) [reply combat/CombatSystem.cpp:1304]: Fixed the comment. Github ate the angle brackets.

Comment (geoffthemedio) [reply combat/CombatSystem.cpp:1304]: > GetWeapons ist not called with shared_ptr<UniverseObject> all the time, but shared_ptr<Ship>.

Not sure of the relevance... pointer-to-derived should be passable as pointer-to-base either way.

My thinking is that passing by value, if not optimized away, would require a copy of the pointer, which involves incrementing the reference count, while passing by const reference would not.

Comment (adrianbroher) [reply combat/CombatSystem.cpp:1304]: You don't pass a pointer-to-derived, but a shared_ptr&lt;T&gt; as shared_ptr&lt;U&gt;&, which are unrelated types.  The fact that T is-a U doesn't matter for the compiler here.

**Comment by geoffthemedio** [comment]
As in https://github.com/freeorion/freeorion/commit/1ac218d51a93641e6367d5cd7130307ce1af9d7d it's currently awkward to iterate over the type-filtered contents of `ObjectMap` (ie. only systems, not all objects), since `all` creates a vector and fills it with objects and returns it, and range-based for doesn't have a template parameter to select a templated `begin` overload.

Do you have a better suggestion than replacing
````
for (auto& system : Objects().all<System>()) {
````
with
````
for (auto sys_it = objects.begin<System>(); sys_it != objects.end<System>(); ++sys_it) {
    auto& system = *sys_it;
````
?

**Comment by adrianbroher** [comment]
The idea was to adopt boost ranges in the next changesets for this.  However to adopt ranges I first got rid of constructs like you just commited.

**Comment by geoffthemedio** [comment]
OK, then I won't do any more and feel free to revert when an alternative is implemented.
